### PR TITLE
Modify J-job and some other scripts for WCOSS2 transition:

### PR DESCRIPTION
**Pull request by DingchenHou-NOAA**
For prod code/data structure changes
1) $COMROOT:
    In its sub-structures $ver replaced prod in each J-job
    $ver and ${ver_xxx} are added to parm/gefs.parm (xxx=gfs/cfs/ec/uk/na/nam)
    In parm/gefs_prep_emissions.parm, $COMROOT lines are moved to J-job
    In scripts/exgefs_forecast.sh, $COMROOT related lines moved to J-job
2) $DCOMROOT: prod is dropped from its sub-structures
3) $GESROOT structure is removed related directories are under $COMOUT
4) $DATAROOT, tmpnwprd is changed to tmp
5) $NWROOT is replaced by $PACKAGEROOT
6) Remove lines with $jlogfile in J-job files and one script file
7) parm/gefs_init.parm is simplified to remove legacy variables

### CODE REVIEW

File: jobs/JGEFS_ATMOS_GEMPAK_META [hunk 1]
```diff
@@ -67,15 +67,16 @@ export FIXgempak=${FIXgempak:-$GEMPAKgefs/fix}
...
+export COMIN=${COMOUT:-${COMROOT}/${NET}/${ver}/${RUN}.${PDY}/$cyc}
+export COMINs=${COMINs:-${COMROOT}/${NET}/${ver}}
+export COMOUT=${COMOUT:-${COMROOT}/${NET}/${ver}/${RUN}.${PDY}/$cyc}
 
-export COMINsgfs=${COMINsgfs:-$(compath.py gfs/prod)}
-export COMINnawips=${COMINnawips:-$(compath.py nawips/prod)}
-export COMINecmwf=${COMINecmwf:-$(compath.py ecmwf/prod/ecmwf)}
-export COMINukmet=${COMINukmet:-$(compath.py nawips/prod/ukmet)}
-export COMINnam=${COMINnam:-$(compath.py nam/prod)}
+export COMINsgfs=${COMINsgfs:-$(compath.py gfs/${ver_gfs})}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GEMPAK_META:75]: Based on the current compath.py on WCOSS2, it should ${envir}/com/gfs/${ver_gfs}, the same as others. However, we can wait for a while to get the final version of the format

File: jobs/JGEFS_ATMOS_GETCFSSST [hunk 2]
```diff
@@ -1,19 +1,19 @@
 #!/bin/ksh
 
-echo "$(date -u) begin ${0}"
+echo "$(date -u) begin ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GETCFSSST:3]: it should be change to "0"

File: jobs/JGEFS_ATMOS_GETCFSSST [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
 
 set -xa
 if [[ ${STRICT:-NO} == "YES" ]]; then
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GETCFSSST:6]: change .sh.file to 0

File: jobs/JGEFS_ATMOS_GETCFSSST [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
-export PS4='$SECONDS + ${0}[$LINENO] '
+export PS4='$SECONDS + $(basename ${.sh.file})[$LINENO] '
 
 ####################################
 # Determine Job Output Name on System
 ####################################
-export jobid=${jobid:-"$job.$PBS_JOBID"}
-export pgmout=${pgmout:-"OUTPUT.${PBS_JOBID}"}
+export jobid=${jobid:-"$job.$LSB_JOBID"}
+export pgmout=${pgmout:-"OUTPUT.${LSB_JOBID}"}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GETCFSSST:16]: LSB_JOBID should be change to PBS_JOBID

File: jobs/JGEFS_ATMOS_GETCFSSST [hunk 5]
```diff
@@ -76,7 +75,7 @@ export err=$?
 if [[ $err == 0 ]]; then
     echo "$job completed normally!"
 else
-    echo "FATAL ERROR in ${0}: $job failed!"
+    echo "FATAL ERROR in ${.sh.file}: $job failed!"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GETCFSSST:78]: .sh.file changes to 0 (ZERO)

File: jobs/JGEFS_ATMOS_GETCFSSST [hunk 6]
```diff
@@ -87,6 +86,6 @@ fi
 cd $DATAROOT
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
 
-echo "$(date -u) end ${0}"
+echo "$(date -u) end ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_GETCFSSST:90]: change .sh.file to 0

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 7]
```diff
@@ -1,19 +1,19 @@
-#!/bin/ksh -l
+#!/bin/ksh
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:1]: Add -l due to the wcoss2's standard

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 8]
```diff
@@ -1,19 +1,19 @@
...
 
-echo "$(date -u) begin ${0}"
+echo "$(date -u) begin ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:3]: change .sh.file to 0

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 9]
```diff
@@ -1,19 +1,19 @@
...
 set -xa
 if [[ ${STRICT:-NO} == "YES" ]]; then
 	# Turn on strict bash error checking
 	set -eu
 fi
-export PS4='$SECONDS + $(basename ${0})[$LINENO] '
+export PS4='$SECONDS + $(basename ${.sh.file})[$LINENO] '
 
 ####################################
 # Determine Job Output Name on System
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:13]: change .sh.file to 0

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
 ####################################
-export jobid=${jobid:-"$job.$PBS_JOBID"}
-export pgmout="OUTPUT.${PBS_JOBID}"
+export jobid=${jobid:-"$job.$LSB_JOBID"}
+export pgmout="OUTPUT.${LSB_JOBID}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:16]: Change LSB_ to PBS

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 11]
```diff
@@ -77,13 +78,11 @@ for config in $configs; do
     . $config_path/$config.parm
     export err=$?
     if [[ $err != 0 ]]; then
-        echo "FATAL ERROR in ${0}: Error while loading parm file $config_path/$config.parm"
-        exit $err
+    	echo "FATAL ERROR in ${.sh.file}: Error while loading parm file $config_path/$config.parm"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:81]: Change .sh.file to 0

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 12]
```diff
@@ -94,7 +93,7 @@ export err=$?
 if [[ $err == 0 ]]; then
     echo "$job completed normally!"
 else
-    echo "FATAL ERROR in ${0}: $job failed!"
+    echo "FATAL ERROR in ${.sh.file}: $job failed!"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:96]: Change .sh.file to 0

File: jobs/JGEFS_ATMOS_INIT_RECENTER [hunk 13]
```diff
@@ -105,6 +104,6 @@ fi
 cd $DATAROOT
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
 
-echo "$(date -u) end ${0}"
+echo "$(date -u) end ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_ATMOS_INIT_RECENTER:107]: Change .sh.file to 0

File: jobs/JGEFS_CHEM_INIT [hunk 14]
```diff
@@ -1,19 +1,19 @@
 #!/bin/ksh
 
-echo "$(date -u) begin ${0}"
+echo "$(date -u) begin ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:3]: Change .sh.file to 0

File: jobs/JGEFS_CHEM_INIT [hunk 15]
```diff
@@ -1,19 +1,19 @@
...
 
 set -xa
 if [[ ${STRICT:-NO} == "YES" ]]; then
 	# Turn on strict bash error checking
 	set -eu
 fi
-export PS4='$SECONDS + $(basename ${0})[$LINENO] '
+export PS4='$SECONDS + $(basename ${.sh.file})[$LINENO] '
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:11]: Change .sh.file to 0

File: jobs/JGEFS_CHEM_INIT [hunk 16]
```diff
@@ -1,19 +1,19 @@
...
 
 ####################################
 # Determine Job Output Name on System
 ####################################
-export jobid=${jobid:-"$job.$PBS_JOBID"}
-export pgmout="OUTPUT.${PBS_JOBID}"
+export jobid=${jobid:-"$job.$LSB_JOBID"}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:15]: LSB to PBS_

File: jobs/JGEFS_CHEM_INIT [hunk 17]
```diff
@@ -26,8 +26,8 @@ cd $DATA
 ######################################
 # Set up compute resources
 ######################################
-export APRUN_CALCINC=${APRUN_CALCINC:-"mpiexec -n 1 -ppn 1"}
-export APRUN_CHGRES=${APRUN_CHGRES:-"mpiexec -n 1 -ppn 1"}
+export APRUN_CALCINC=${APRUN_CALCINC:-"mpirun -n 1"}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:29]: Change "mpirun -n 1" to "mpiexec -n 1 -ppn 1" for WCOSS2

File: jobs/JGEFS_CHEM_INIT [hunk 18]
```diff
@@ -80,13 +81,11 @@ for config in $configs; do
     . $config_path/$config.parm
     export err=$?
     if [[ $err != 0 ]]; then
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:83]: Change .sh.file to 0

File: jobs/JGEFS_CHEM_INIT [hunk 19]
```diff
@@ -97,7 +96,7 @@ export err=$?
 if [[ $err == 0 ]]; then
     echo "$job completed normally!"
 else
-    echo "FATAL ERROR in ${0}: $job failed!"
+    echo "FATAL ERROR in ${.sh.file}: $job failed!"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:99]: .sh.file -> 0

File: jobs/JGEFS_CHEM_INIT [hunk 20]
```diff
@@ -108,6 +107,6 @@ fi
 cd $DATAROOT
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
 
-echo "$(date -u) end ${0}"
+echo "$(date -u) end ${.sh.file}"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_INIT:111]: .sh.file -> 0

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 21]
```diff
@@ -1,19 +1,19 @@
 #!/bin/bash
 
-echo "$(date -u) begin $(basename ${0})"
+echo "$(date -u) begin $(basename $BASH_SOURCE)"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:4]: Change BASH_SOURCE to 0

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 22]
```diff
@@ -1,19 +1,19 @@
...
 
 set -xa
 if [[ ${STRICT:-NO} == "YES" ]]; then
 	# Turn on strict bash error checking
 	set -eu
 fi
-export PS4='$SECONDS + $(basename ${0}))[$LINENO] '
+export PS4='$SECONDS + $(basename $BASH_SOURCE))[$LINENO] '
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:10]: Change BASH_SOURCE to 0

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 23]
```diff
@@ -1,19 +1,19 @@
...
 
 ####################################
 # Determine Job Output Name on System
 ####################################
-export jobid=${jobid:-"$job.$PBS_JOBID"}
-export pgmout="OUTPUT.${PBS_JOBID}"
+export jobid=${jobid:-"$job.$LSB_JOBID"}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:15]: Change LSB to PBS

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 24]
```diff
@@ -70,12 +71,17 @@ for config in $configs; do
     . $config_path/$config.parm
     export err=$?
     if [[ $err != 0 ]]; then
-        echo "FATAL ERROR in $(basename ${0}): Error while loading parm file $config_path/$config.parm"
-        exit $err
+    	echo "FATAL ERROR in $(basename $BASH_SOURCE): Error while loading parm file $config_path/$config.parm"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:74]: Change BASH_SOURCE to 0

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 25]
```diff
@@ -87,7 +93,7 @@ export err=$?
 if [[ $err == 0 ]]; then
     echo "$job completed normally!"
 else
-    echo "FATAL ERROR in $(basename ${0}): $job failed!"
+    echo "FATAL ERROR in $(basename $BASH_SOURCE): $job failed!"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:96]: Change BASH_SOURCE to 0

File: jobs/JGEFS_CHEM_PREP_EMISSIONS [hunk 26]
```diff
@@ -98,6 +104,6 @@ fi
 cd $DATAROOT
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
 
-echo "$(date -u) end $(basename ${0})"
+echo "$(date -u) end $(basename $BASH_SOURCE)"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_CHEM_PREP_EMISSIONS:107]: Change BASH_SOURCE to 0

File: jobs/JGEFS_WAVE_INIT [hunk 27]
```diff
@@ -1,27 +1,27 @@
 #!/bin/bash
 
-echo "$(date -u) begin $(basename ${0})"
+echo "$(date -u) begin $(basename $BASH_SOURCE)"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:3]: Change BASH_SOURCE to 0

File: jobs/JGEFS_WAVE_INIT [hunk 28]
```diff
@@ -1,27 +1,27 @@
...
 fi
-export PS4='$SECONDS + $(basename ${0})[$LINENO] '
+export PS4='$SECONDS + $(basename $BASH_SOURCE)[$LINENO] '
 
 ####################################
 # Determine Job Output Name on System
 ####################################
-export jobid=${jobid:-"$job.$PBS_JOBID"}
-export pgmout="OUTPUT.${PBS_JOBID}"
+export jobid=${jobid:-"$job.$LSB_JOBID"}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:15]: LSB to PBS

File: jobs/JGEFS_WAVE_INIT [hunk 29]
```diff
@@ -1,27 +1,27 @@
...
+export pgmout="OUTPUT.${LSB_JOBID}"
 export pgmerr=errfile
 
 ######################################
 # Set up compute resources
 ######################################
+export wavempexec=${wavempexec:-"mpirun -n"}
+export wave_mpmd=${wave_mpmd:-"cfp"}
 export NTASKS=${total_tasks:-24}
-export wavempexec=${wavempexec:-"mpiexec -cpu-bind core"} # -n "}
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:23]: Keep this change for WCOSS2

File: jobs/JGEFS_WAVE_INIT [hunk 30]
```diff
@@ -82,7 +83,7 @@ for config in $configs; do
     . $config_path/$config.parm
     export err=$?
     if [[ $err != 0 ]]; then
-    	echo "FATAL ERROR in $(basename ${0}): Error while loading parm file $config_path/$config.parm"
+    	echo "FATAL ERROR in $(basename $BASH_SOURCE): Error while loading parm file $config_path/$config.parm"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:86]: BASH_SOURCE -> 0

File: jobs/JGEFS_WAVE_INIT [hunk 31]
```diff
@@ -105,7 +105,7 @@ export err=$?
 if [[ $err == 0 ]]; then
     echo "$job completed normally!"
 else
-    echo "FATAL ERROR in $(basename ${0}): $job failed!"
+    echo "FATAL ERROR in $(basename $BASH_SOURCE): $job failed!"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:108]: BASH_SOURCE -> 0

File: jobs/JGEFS_WAVE_INIT [hunk 32]
```diff
@@ -116,6 +116,6 @@ fi
 cd $DATAROOT
 if [ ${KEEPDATA:-NO} = NO ] ; then rm -rf $DATA ; fi
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_INIT:118]: BASH_SOURCE -> 0

File: jobs/JGEFS_WAVE_PREP [hunk 33]
```diff
@@ -1,13 +1,13 @@
 #!/bin/bash
 
-echo "$(date -u) begin $(basename ${0})"
+echo "$(date -u) begin $(basename $BASH_SOURCE)"
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_PREP:3]: BASH_SOURCE -> 0

File: jobs/JGEFS_WAVE_STAT [hunk 34]
```diff
@@ -58,17 +58,18 @@ source PDY
...
+export HOMEgefs=${HOMEgefs:-${PACKAGEROOT}/gefs.${gefs_ver}}
 export PARMgefs=${PARMgefs:-$HOMEgefs/parm}
 export FIXwave=${FIXwave:-$HOMEgefs/fix/fix_wave}
 
 ##############################################
 # Define COM directories
 ##############################################
 export COMPONENT=wave
+ver=${ver:-$(echo ${gefs_ver}|cut -c2-3)}
 # Have to include $COMPONENT in COM paths for downstream scripts
```
Comment (XianwuXue-NOAA) [inline jobs/JGEFS_WAVE_STAT:70]: Why change the original COMPONENT to wave?

File: parm/gefs.parm [hunk 35]
```diff
@@ -1,4 +1,4 @@
-#!/bin/ksh -l
+#!/bin/ksh
```
Comment (XianwuXue-NOAA) [inline parm/gefs.parm:1]: Add -l for wcoss2 standard

**Review by XianwuXue-NOAA — CHANGES_REQUESTED** [review]
Based on WCOSS2 standard, therefore, I did some changes and tested on Acorn. Due to legacy ksh on Cactus (acorn has updated the ksh, so some issues have solved), we must change .sh.file or BASHSOURCE to 0 and LSB to PBS. So please keep my changes for the initial tasks.

**Comment by DingchenHou-NOAA** [comment]
My understanding is that my changes will be merged into your branch without
conflict. Your changes will not be affected.
I have noted what you did.

**Comment by XianwuXue-NOAA** [comment]
I think this is because your changes is newer than my changes, so it will have such conflict. Usually, I will merge the upper repo before I submit PR/code-review, during the merge, I will solve the conflict. Then the PR will keep the newer changes both yours and mines

**Comment by DingchenHou-NOAA** [comment]
I see the problem. Will drop this PR and create a new one later.

**Comment by XianwuXue-NOAA** [comment]
You don't need drop this PR, you can continue working on your branch and your changes will reflect to this PR. The PR is different with the previous code-review in vlab. The PR will show all your changes on your branch until merged or canceled the PR, even the reviewer has approved it.

**Pull request closed without merging**